modules/thermal_hydraulics/src/materials/ADWallFrictionChengMaterial.C

```diff
@@ -0,0 +1,406 @@
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ADWallFrictionChengMaterial.h"
+#include "WallFrictionModels.h"
+#include "SinglePhaseFluidProperties.h"
+#include "FlowModelSinglePhase.h"
+#include "Numerics.h"
+
+registerMooseObject("ThermalHydraulicsApp", ADWallFrictionChengMaterial);
+
+InputParameters
+ADWallFrictionChengMaterial::validParams()
+{
+  InputParameters params = Material::validParams();
+  params.addRequiredParam<MaterialPropertyName>("f_D", "Darcy friction factor material property");
+  params.addParam<MaterialPropertyName>(
+      "rho", FlowModelSinglePhase::DENSITY, "Density of the liquid");
+  params.addParam<MaterialPropertyName>("vel", FlowModelSinglePhase::VELOCITY, "Fluid velocity");
+  params.addParam<MaterialPropertyName>(
+      "D_h", FlowModelSinglePhase::HYDRAULIC_DIAMETER, "Hydraulic diameter");
+  params.addParam<MaterialPropertyName>(
+      "mu", FlowModelSinglePhase::DYNAMIC_VISCOSITY, "Dynamic viscosity of the fluid");
+  params.addRequiredParam<Real>(
+      "PoD", "The Pitch-to-diameter ratio value being assigned into the property");
+  MooseEnum bundle_array("SQUARE HEXAGONAL");
+  params.addRequiredParam<MooseEnum>(
+      "bundle_array", bundle_array, "The type of the rod bundle array");
+  MooseEnum subchannel_type("INTERIOR EDGE CORNER");
+  params.addRequiredParam<MooseEnum>(
+      "subchannel_type", subchannel_type, "The type of subchannel to be considered");
+  params.addClassDescription(
+      "Computes wall friction factor using the Cheng-Todreas correlation for interior channels.");
+  return params;
+}
+
+ADWallFrictionChengMaterial::ADWallFrictionChengMaterial(const InputParameters & parameters)
+  : Material(parameters),
+    _f_D_name(getParam<MaterialPropertyName>("f_D")),
+    _f_D(declareADProperty<Real>(_f_D_name)),
+    _rho(getADMaterialProperty<Real>("rho")),
+    _vel(getADMaterialProperty<Real>("vel")),
+    _D_h(getADMaterialProperty<Real>("D_h")),
+    _mu(getADMaterialProperty<Real>("mu")),
+    _PoD(getParam<Real>("PoD")),
+    _bundle_array(getParam<MooseEnum>("bundle_array").getEnum<Bundle_array>()),
+    _subchannel_type(getParam<MooseEnum>("subchannel_type").getEnum<Subchannel_type>())
+
+{
+}
+
+void
+ADWallFrictionChengMaterial::computeQpProperties()
+{
+  ADReal Re = THM::Reynolds(1, _rho[_qp], _vel[_qp], _D_h[_qp], _mu[_qp]);
+  if (_PoD > 1.5)
+  {
+    mooseDoOnce(mooseWarning(
+        "The Cheng-Todreas correlation for the friction factor is valid when P/D is between 1.0 "
+        "and 1.5. Be aware that using values out of this range may lead to "
+        "significant errors in your results!"));
+  }
+  else if (_PoD < 1.0)
+  {
+    mooseError("Your Pitch-to-Diameter ratio must be equal of above 1.0.");
+  }
+  // The Pitch-to-Diameter ratio (PoD) cannot be smaller than 1.0.
```

Comment:
Make the parameter range-checked instead of this. You can then delete the corresponding error test.